### PR TITLE
support remote rm api for ppo and ppo ray

### PR DESCRIPTION
**Pull request by catqaq**
Now everyone can easily use remote RM (such as Nemotron 340B RM) to train the PPO model. Since this feature has been delayed for a long time, the old version of remote RM PPO without Ray is implemented based on (openrlhf0.2.6), and the new version with remote RM PPO with Ray is implemented based on (openrlhf0.3.2), which has some conflicts with the latest main branch. Today, I manually migrated it to the latest main code to resolve the conflicts. However, due to limited energy and resources, I cannot test the latest code in the near future, and I hope everyone can help test it.

1. old version: https://github.com/OpenLLMAI/OpenRLHF/pull/341
2. test in v0.2.6/0.3.2
3. test rm api: 7b/13b/100b+

### CODE REVIEW

**Comment by hijkzzz** [comment]
I will review this MR when I have bandwidth. Thx.

File: openrlhf/cli/train_ppo_ray.py [hunk 1]
```diff
@@ -110,27 +110,34 @@ def train(args):
...
-    for reward_model, reward_pretrain in zip(reward_models, reward_pretrains):
-        refs.extend(reward_model.async_init_model_from_pretrained(strategy, reward_pretrain))
+    if not args.remote_rm_url:
+        for reward_model, reward_pretrain in zip(reward_models, reward_pretrains):
+            refs.extend(reward_model.async_init_model_from_pretrained(strategy, reward_pretrain))
 
     # init vLLM engine for text generation
     vllm_engines = None
-    if args.vllm_num_engines is not None:
+    if args.vllm_num_engines > 0:
```
Comment (hijkzzz) [inline openrlhf/cli/train_ppo_ray.py:140]: if args.vllm_num_engines is not None and args.vllm_num_engines > 0:

Comment (hijkzzz) [inline openrlhf/cli/train_ppo_ray.py:140]: More safe:
`if args.vllm_num_engines is not None and args.vllm_num_engines > 0:`

**Comment by catqaq** [comment]
1."I suggest removing the remote reference model, as it is redundant with Ray (https://openrlhf.readthedocs.io/en/latest/multi-node.html). Keep the remote reward model." 
In fact, the remote RM feature has essentially been removed, with only some experimental APIs retained under PPO without ray. By default, these are already commented out in the shell script and are only reserved for special experimental purposes and future work.

2."It would be best if the remote RM is compatible with the vLLM server (OpenAI entrypoints?) or Ollama servers."
For the time being, we don’t need to worry about what starts the remote RM. Of course, in the future, we can provide scripts and tutorials for launching services via vLLM or Ollama server.

**Comment by hijkzzz** [comment]
1. In fact, the remote ref feature has essentially been removed. Just remove the unused codes.
2. How to init critic model is a new issue with remote RM.

**Comment by catqaq** [comment]
tested in v0.2.6/0.3.2/0.3.6.
main brach is so active. qaq 
-> there are some conflicts with the latest version 0.3.7, which might be resolved tomorrow.

**Comment by catqaq** [comment]
remote rm is ready. https://github.com/OpenRLHF/OpenRLHF/pull/378
The remaining RM inference service acceleration is left to experts in the field of inference.


> 1. "In fact, the remote ref feature has essentially been removed." **Just remove the unused codes.**
> 2. How to init critic model is a new issue with remote RM. (Use actor model weights?)
> 3. Please provide the scripts and tutorials for launching server via vLLM or Ollama server in examples.

**Comment by hijkzzz** [comment]
The bugs in this MR are numerous and need further fixing. @xiaoxigua999

**Comment by catqaq** [comment]
ray remote rm fn was ready.

**Comment by catqaq** [comment]
For the RM server, no mature solution has been found for the time being, and it may require experts in reasoning to assist. It is suggested to consider it as a separate feature.

https://github.com/vllm-project/vllm/issues/6620
https://github.com/hiyouga/LLaMA-Factory/issues/4834
https://github.com/hiyouga/LLaMA-Factory/issues/1543
https://github.com/hiyouga/LLaMA-Factory/pull/1954

**Pull request closed without merging**